arch-design-questions.md

```diff
@@ -0,0 +1,173 @@
+
+We anticipate that this will be a fairly common pattern. To simplify the experience for the user each
+language will have some form of a **session manager**.
+
+This client is a higher-level client that internally manages the connection on behalf of the user, does
+manage state and is distinct from the [single use clients](single-use-clients).
+
+Existing solutions:
+
+- JS: [`SessionManager` (has not been reviewed for production-readiness)](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/servicebus/service-bus/src/session/sessionManager.ts)
+- .NET: [`ServiceBusProcessor`](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/servicebus/Azure.Messaging.ServiceBus/src/Processor/ServiceBusProcessorClient.cs)
+- Java/Python - no solution currently.
+
+# Connection sharing
+
+Connection sharing is tricky for some of the SDKs due to the lack of a top-level object. This means
+that individual clients that are created need to have some method for receiving a "shared" object 
+that can contain an AMQP connection (or equivalent).
+
+There are a few proposals in play, which we will revisit after preview.1:
+
+## Connection sharing via shared object
+
+.NET currently has the ServiceBusConnection class to facilitate connection sharing. 
+The constructor for the individual clients take either the service bus connection object 
+_or_ the individual connection parameters.
+
+This solution allows for the simple path (creating directly via connection string/token credential) while
+also allowing a more savvy user to share connections. 
+
+The downside to this approach (mostly in implementation) is that the same overloads that exist for
+both the sender _and_ receiver must be present on the ServiceBusConnection object. 
+
+As this is considered an advanced scenario (and thus useful for advanced users) this downside 
+is trivial to work around.
+
+## Connection sharing via top level factory
+
+Python, with it's top level ServiceBusClient, is in connection-sharing mode by _default_. 
+
+```python
+client = new ServiceBusClient()
+queueClient = client.createQueueClient()
+topicClient = client.createTopicClient()
+```
+
+All instances spawned from the `client` above would share an AMQP connection implicitly
+without any additional work to the user.
+
+The downsides of this approach is that having a top level object is considered non-idiomatic 
+for the other languages, which makes this a non-universal solution.
+
+## Connection sharing by requiring a shared object
+
+One possibility to address the constructor-pattern duplication of  ["connection sharing via shared object"](#connection-sharing-via-shared-object)
+is to _require_ the `ServiceBusConnection` class as an argument. 
+
+This would mean the individual clients could not be instantiated without first instantiating a connection class.
+
+The downsides to this approach are that users will be forced (even in the "common" case) to create two objects,
+rather than one. 
```

Comment:
@HarshaNalluru's proposal:

```js
// ServiceBusConnection
const conn =  new ServiceBusConnection(connectionString, options);
or
const conn =  new ServiceBusConnection(fullyQualifiedNamespace, credential, options);

// ServiceBusReceiverClient
const client = new ServiceBusReceiverClient(conn, queue-name, options)
or
const client = new ServiceBusReceiverClient(conn, topic-name, subscription-name, options)

client.close()
conn.close()
```

@richardpark-msft's concerns:
> This would mean the individual clients could not be instantiated without first instantiating a connection class.
>The downsides to this approach are that users will be forced (even in the "common" case) to create two objects, rather than one. 

**My proposal**:

```js
// ServiceBusConnection
const serviceBusConnection =  new ServiceBusConnection(connectionString, options);
or
const serviceBusConnection =  new ServiceBusConnection(fullyQualifiedNamespace, credential, options);

// ServiceBusReceiverClient
const client = new ServiceBusReceiverClient(serviceBusConnection, queue-name, options);
or
const client = new ServiceBusReceiverClient(serviceBusConnection, topic-name, subscription-name, options);
or 
const client = new ServiceBusReceiverClient(connectionString);
or
const client = new ServiceBusReceiverClient(connectionString, queueOrSubscriptionName, options);
or
const client = new ServiceBusReceiverClient(connectionString, topicName, subscriptionName, options);
or
const client = new ServiceBusReceiverClient(fullyQualifiedNamespace, queueName, credential, options);
or
const client = new ServiceBusReceiverClient(fullyQualifiedNamespace, topicName, subscriptionName, credential, options);

client.close()
conn.close()
```
Some advantages of this approach:
* Users can still use other client overloads.
* If users don't want to share the connection, with this approach they can instantiate client without instantiating a connection class
* Now users won't be forced to create 2 objects, rather than one.


If we consider this approach, then do we need one more client constructor without entity name, in case user passes `queue-conn-string` or `topic-conn-string` in serviceBusConnection constructor?

```js
const client = new ServiceBusReceiverClient(serviceBusConnection, options);
```

Reply:
+1 for Shivangi's proposal.

The only mention that I'll make is that we've been asked to order the parameters such that those which are consistent between constructors come first and those that change next, followed by the options at the end.   So in this case, something like:

```csharp
// ServiceBusConnection
const serviceBusConnection =  new ServiceBusConnection(connectionString, options);
or
const serviceBusConnection =  new ServiceBusConnection(fullyQualifiedNamespace, credential, options);

// ServiceBusReceiverClient
const client = new ServiceBusReceiverClient(queue-name, serviceBusConnection, options);
or
const client = new ServiceBusReceiverClient(topic-name, subscription-name, serviceBusConnection, options);
or 
const client = new ServiceBusReceiverClient(connectionString);
or
const client = new ServiceBusReceiverClient(connectionString, queueOrSubscriptionName, options);
or
const client = new ServiceBusReceiverClient(connectionString, topicName, subscriptionName, options);
or
const client = new ServiceBusReceiverClient(fullyQualifiedNamespace, queueName, credential, options);
or
const client = new ServiceBusReceiverClient(fullyQualifiedNamespace, topicName, subscriptionName, credential, options);
```

Reply:
Shivangi's proposal is what we are proposing for .NET as well.

Reply:
> If we consider this approach, then do we need one more client constructor without entity name, in case user passes queue-conn-string or topic-conn-string in serviceBusConnection constructor?

I would vote against doing this and keeping entities out of the Connection.

Reply:
I think I'm digging @ShivangiReja 's proposal as well with one small modification:

- Make it so ServiceBusConnection only takes the "namespace" parameters, not other constructor parameters. So:

ServiceBusConnection(connection string or entity connection string, <no options parameter here>)
ServiceBusConnection(fqdn, token credential)

So going along with what @JoshLove-msft is saying - keep entity name out of the ServiceBusConnection and treat it as a synonym for the connection string. We can look at the client constructors as being how the user drills down to a topic/queue/subscription.

There's still some duplication but at least  now it just looks like we swap out one parameter for another (and they're both named "conection-ish" something) which feels a bit better to my mind.